### PR TITLE
fix/Notary request tracking

### LINKED ISSUE #2315: Notary request tracking is incorrect
## Expected Behavior
Notary requests should be tracked by their actions and main transaction hash.

## Current Behavior
Looking at signatures to determine if it's a new notary request to handle is just plain wrong (see notary preparator code in #2310).

## Possible Solution
Rework this mechanism to improve reliability.

### PR DESCRIPTION
**Pull request by carpawell**
Closes #2315

### CODE REVIEW

**Comment by carpawell** [comment]
@roman-khimov, TBH, yesterday I figured out that it was not a problem at all. Yes, we parse every notary request and are ready to sign each of them **but** not every method has a handler and a parser for it: e.g. our beloved `NewEpoch` event is listened as a simple notification cause netmap processor has a [parser](https://github.com/nspcc-dev/neofs-node/blob/0168dc5df9dff0a2a253b7dd66a81319f192cad1/pkg/innerring/processors/netmap/processor.go#L169-L172) and a [handler](https://github.com/nspcc-dev/neofs-node/blob/0168dc5df9dff0a2a253b7dd66a81319f192cad1/pkg/innerring/processors/netmap/processor.go#L184-L187) for it. It is ready to update its internal state and be sure that everyone thinks that a `NewEpoch` has happened cause it believes the contract. But if another IR sends notary requests it won't be handled cause IR does not have `NewEpoch` _notary_ [parser](https://github.com/nspcc-dev/neofs-node/blob/0168dc5df9dff0a2a253b7dd66a81319f192cad1/pkg/innerring/processors/netmap/processor.go#L203-L211) and _notary_ [handler](https://github.com/nspcc-dev/neofs-node/blob/0168dc5df9dff0a2a253b7dd66a81319f192cad1/pkg/innerring/processors/netmap/processor.go#L227-L235) for it.

But why have I added the last commit then? I still think that `morph` refactor is somewhere near and we will drop (simplify) `Listener`. Also, I think notary events are missing contract side checks (the `netmap` contract will not allow ticking epoch if there are less than 5 IR requested a new epoch; **but** IR would tick epoch if there are a hander for it and a corresponding notary request has been parsed) so I think it should be a notary preparator (`preparator` now) who must skip unnecessary requests not a missing handler.

**Review by roman-khimov — APPROVED** [review]

File: pkg/morph/event/notary_preparator.go [hunk 1]
```diff
@@ -43,6 +44,9 @@ var (
 
 	// ErrMainTXExpired is returned if received fallback TX is already valid.
 	ErrMainTXExpired = errors.New("received main tx has expired")
+
+	// ErrUnknownEvent is returned if main transaction is not expected.
+	ErrUnknownEvent = errors.New("received notary request is not allowed")
```
Comment (AnnaShaleva) [inline pkg/morph/event/notary_preparator.go:49]: Don't we need to add a special case for this event to `func (l *listener) parseAndHandleNotary(nr *result.NotaryRequestEvent) {`? We may probably safely skip logging for these events as they are unexpected and unwanted (currently `could not prepare and validate notary event` warning will be logged).

**Review by AnnaShaleva — COMMENTED** [review]
Otherwise LGTM.

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]
We've discussed and came to the solution that transctions cache isn't needed and can be removed. The only purpose of this cache is to filter out notary requests that were sent by the alphabet node itself. Instead of this we can add a check for notary request: the second signer of fallback transaction shouldn't be the committee member.

Disadvantage of this approach is that we have to always keep the list of committee members up-to-date. It can be reached with RPC-based cache (but then a lot of `getcommittee` RPC requests are expected) or with cache that is subscribed to the committee updates (we need to add some code to our cache than).

@roman-khimov, what do you think?

**Comment by roman-khimov** [comment]
Do you mean `alreadyHandledTXs`?

> sent by the alphabet node itself
> the second signer of fallback transaction shouldn't be the committee member

So is it a node key that is a problem of is it a whole committee that is a problem? What if SN uses the same key as some IR node?

Hi, neo-project/neo#2763, anyway.

**Comment by AnnaShaleva** [comment]
> Do you mean alreadyHandledTXs?

Yes.

> What if SN uses the same key as some IR node?

But is it possible? I thought that SN and IR are always different antities with different accounts. If not, then the described way isn't workable.

**Comment by roman-khimov** [comment]
Nothing technically prevents SN from using the same key as IR.

File: pkg/morph/event/notary_preparator.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/morph/event/notary_preparator.go:49]: @carpawell, could you push the update?

Comment (carpawell) [reply pkg/morph/event/notary_preparator.go:49]: sure, that was an error, i thout about extending `switch` but forgot about it

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**